### PR TITLE
Improved the performance of the Arr::set() method by ~50%

### PR DESCRIPTION
**Pull request by lindelius**
**Closed due to side-effects not wanted for a minor release.**

### CODE REVIEW

**Review by carusogabriel — CHANGES_REQUESTED** [review]
@lindelius Benchmarks?

**Comment by antonkomarev** [comment]
@lindelius that's a good improvement. It's a good practice to write benchmarks in comments of optimization PR.

File: src/Illuminate/Support/Arr.php [hunk 1]
```diff
@@ -517,24 +517,20 @@ public static function set(&$array, $key, $value)
...
-            // values at the correct depth. Then we'll keep digging into the array.
-            if (! isset($array[$key]) || ! is_array($array[$key])) {
-                $array[$key] = [];
+        $segments = explode('.', $key);
+
+        foreach ($segments as $segment) {
+            // If the key doesn't exist at this depth, we'll just create an
+            // empty array to hold the next value, allowing us to create the
+            // arrays to hold the final value at the correct depth.
+            if (isset($array) && ! is_array($array)) {
```
Comment (laurencei) [inline src/Illuminate/Support/Arr.php]: Why does this test even need to be inside the loop? Just set the `$array = []` outside of the `foreach` - then remove this whole section.

**Comment by lindelius** [comment]
@antonkomarev Very true. I'm running PHP 7.2.7 on Ubuntu 16.04 and I used the following script in an untouched Laravel 5.7 installation

```php
<?php

require __DIR__ . '/vendor/autoload.php';

$key = '';

for ($l = 1; $l <= 10; $l++) {
    $key   = ltrim($key . '.level' . $l, '.');
    $start = microtime(true);

    for ($i = 0; $i < 1000000; $i++) {
        $array = [];
        Illuminate\Support\Arr::set($array, $key, 'value_' . $i);
    }

    echo round((microtime(true) - $start) / 1000, 6) . "\n";
}
```

The numbers are pretty much identical for setting values just one level deep (as one might expect), but the performance difference increases with the number of levels.

The results (seconds average per 1,000 iterations) were the following

| levels | before | after |
| :--: | :--------: | :---------: |
|  1 | 0.00031 | 0.00031 |
|  2 | 0.00052 | 0.00044 |
|  3 | 0.00075 | 0.00056 |
|  4 | 0.00098 | 0.00069 |
|  5 | 0.00116 | 0.00082 |
|  6 | 0.00137 | 0.00091 |
|  7 | 0.00158 | 0.00102 |
|  8 | 0.00181 | 0.00114 |
|  9 | 0.00207 | 0.00129 |
| 10 | 0.00241 | 0.00138 |

File: src/Illuminate/Support/Arr.php (hunk 1 above)
Comment (lindelius) [reply src/Illuminate/Support/Arr.php]: Because we need to check this for every level of the (possibly) nested array.

Comment (lindelius) [reply src/Illuminate/Support/Arr.php]: Since it's passed by reference it only overwrites the current level it's at.

**Comment by lindelius** [comment]
Since the unit tests doesn't cover this and since I've never used the helper method in that way, I didn't realize until now that I was actually messing up the return value. However, I just had a look through the code base, as well as all Laravel projects that I've got access to, and there's just a single place where the return value is actually used, which is in the `DatabaseSessionHandler` class.

The proposed changes would therefore be BC breaking as the return value is slightly different, and it would require a minor change on 1 line in the previously mentioned class. If it were up to me, I'd of course make the change, but I'm not the one to make this decision.

**Pull request closed without merging**